lib/reda/importers/mpt_das1.py

```diff
@@ -0,0 +1,621 @@
+                # phase
+                data_new.loc[row_idx, 'rpha'] = float(data.iloc[row_idx, idx*6
+                                                                + 6
+                                                                + tx_out_skip*fskip_count[row_idx]])
+                # current in mA
+                data_new.loc[row_idx, 'I'] = float(data.iloc[row_idx, idx*6
+                                                             + 8
+                                                             + tx_out_skip*fskip_count[row_idx]])
+                # devR
+                data_new.loc[row_idx, 'dr'] = float(data.iloc[row_idx, idx*6
+                                                              + 5
+                                                              + tx_out_skip*fskip_count[row_idx]])
+                # devPhi
+                data_new.loc[row_idx, 'drpha'] = float(data.iloc[row_idx, idx*6
+                                                                 + 7
+                                                                 + tx_out_skip*fskip_count[row_idx]])
+                # array to check for error
+                dout_r[row_idx, idx] = float(data.iloc[row_idx, idx*6
+                                                       + 4
+                                                       + tx_out_skip*fskip_count[row_idx]])
+
+        else:
+
+            dout_r[:, idx] = np.array(data.iloc[:, idx*6 + 4])
+            data_new['r'] = np.array(data.iloc[:, idx*6 + 4]).astype('float')  # resistance
+            data_new['rpha'] = np.array(data.iloc[:, idx*6 + 6]).astype('float')  # phase
+            data_new['I'] = np.array(data.iloc[:, idx*6 + 8]).astype('float')  # current in mA
+            data_new['dr'] = np.array(data.iloc[:, idx*6 + 5]).astype('float')  # devR
+            data_new['drpha'] = np.array(data.iloc[:, idx*6 + 7]).astype('float')  # devPhi
+
+        # check for quadrupoles containing nans (because of error)
+        nan_index = np.where(np.isnan(dout_r[:, idx]) == 1)[0]
+        # set the skip count
+        fskip_count[nan_index] = fskip_count[nan_index] + 1
+        fskip_count.astype(int)
+
+        data_fin = data_fin.append(data_new, ignore_index=True, sort=False)
+
+    # compute Zt
+    data_fin['Zt'] = data_fin['r'] * np.exp(data_fin['rpha'] * 1j / 1000.0)
+
+    start = len(frequencies)-1
+    for row_idx in range(len(data)):
+        data_new.at[row_idx, 'datetime'] = data.iloc[row_idx,
+                                                     start*6 + 11 + tx_out_skip*fskip_count[row_idx]]
+
+    datetime_series = pd.to_datetime(
+        data_new['datetime'], format='%Y%m%d_%H%M%S', errors='ignore')
+    datetime_stack = datetime_series.append([datetime_series]*(num_freqs-1))
+    datetime_stack_reindex = datetime_stack.reset_index()
+    data_fin['datetime'] = datetime_stack_reindex['datetime']
+
+    return data_fin, None, None
+
+
+def import_das1(filename, **kwargs):
```

Comment:
without detailed review, this function seems to copy&paste all the previous functions. This is a lot of code duplication.

What about writing one function that decides on the type of the result file, and then using the previously added functions to import the data.

I don't think its important if the file is opened twice (once for detection, once for data import). Otherwise StringIO could help.

Reply:
Fixed in the reviewed code.